autograder-core/src/main/java/de/firemage/autograder/core/Uses.java

```diff
@@ -0,0 +1,89 @@
+package de.firemage.autograder.core;
+
+import de.firemage.autograder.core.integrated.SpoonUtil;
+import spoon.processing.AbstractProcessor;
+import spoon.reflect.CtModel;
+import spoon.reflect.code.CtVariableAccess;
+import spoon.reflect.declaration.CtConstructor;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtNamedElement;
+import spoon.reflect.declaration.CtParameter;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.declaration.CtTypeMember;
+import spoon.reflect.declaration.CtVariable;
+import spoon.reflect.reference.CtVariableReference;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+public class Uses {
```

Comment:
I am unhappy with the API of this class. It feels more like a hack and makes extensions difficult to implement.

You wrote in the PR that you were able to achieve some performance improvements? I guess they are the result of caching and not from your implementation of the searching in the model?

To be honest, I think it would make more sense to take my `UsesFinder` class and adapting it with your code, so it caches the queries.

Ideally, the caching would be separate from the `UsesFinder`. Maybe a class around `Map<IdentityKey<CtElement>, List<CtElement>> cache` which is stored in the `CodeModel`. (Ideally, the cache would be global, so we do not have to pass the `CodeModel` everywhere we want to search for a use).

And then `UsesFinder` would either use the cached results or compute them and update the cache.

----

I have some ideas about caching, but not sure how much they influence performance:
- The biggest speedup comes from narrowing down the search scope. For example, a `CtParameter` can only be used in the `CtExecutable` that declares it. Similarly, this can be done for other elements like local variables, things that are private or only default visibility. The largest search scope would be for public `CtTypeMember` that could be used anywhere. Maybe those could be sped up by initially iterating through all `CtCompilationUnit` and caching which compilation unit imports which `CtType`. (The only problem here is that one might not import the type, but use a qualified name in the code)
- Maybe use a dedicated caching library like https://github.com/ben-manes/caffeine
- If I remember correctly, I was unable to get caching working, because the spoon model is mutable? Not sure what the problem was...

Reply:
No, the performance gains are coming from the search strategy. Computing uses per e.g. variable always has that O(n^2) behavior that I mentioned above. My implementation basiacally reverses the search - we look for any use of any variable, and record a use of that variable. That uses the fact that Spoon is quite performant at giving you the declaration of an element, but filtering or querying the model and thereby visiting every element is not. I'll provide some numbers later today.